### DISCUSSION TITLE
N64 carts sharing same serial but not same rom size

### DISCUSSION THREAD

**Discussion by PsyK0p4T** [sanni/cartreader#383] [Ideas & Suggestions]
Hi Sanni.
Just saw your recent changes to sd txt files, for rom renaming purposes. That's a cool addition :)
Can't speak for all systems, but for n64 we will maybe have a problem in some very rare cases. Example : 

Star Wars - Shadows of the Empire (USA).z64
3C0837B3,NSWE,12,5
Star Wars - Shadows of the Empire (USA) (Rev 1).z64
B0540688,NSWE,12,5
Star Wars - Shadows of the Empire (USA) (Rev 2).z64
E8727549,NSWE,12,5
Star Wars - Shadows of the Empire (USA) (Aftermarket) (Unl).z64
776A069B,NSWE,16,5

Here, all the retail version of the game are 12MB. So I guess once the cartridge will be inserted, cart reader will first look at the game serial, then search for its size in txt, dump it, compare with known CRCs, and rename rom accordingly.
Aftermarket cart (Limit Run Games edition) also shares the same serial, but its rom size is different. Not sure how the cart reader will act... overdump all "NSWE" retail carts ? or underdump the aftermarket one ?

Not sure about what is the best way to deal with this. Simplicity would be to just avoid these exotic dump (protos, betas, aftermarket...) in txt. 
Or maybe modify a bit the method to identify a game. Not only by its serial but more like a couple of "game serial" + "rom version", wich are both readable in cart header. Then by checking the txt we would expect a precise size and CRC for each unique couple ?

Once again, this case is extremly rare (unique ?)

**Comment by sanni** [comment]
Good catch. I did not think of that. In the SNES code we use the internal rom checksum instead of the game code to look up the database. Probably ran into a similar problem there back then.

**Comment by PsyK0p4T** [comment]
Hi Sanni. I'm digging up this thread again, cause we have the same behavior on GBA.
Example : 

```
Metroid - Zero Mission (USA) (Beta).gba
922C4A10,BMXE,16,3

Metroid - Zero Mission (USA) (Virtual Console).gba
8825294A,BMXE,08,3

Metroid - Zero Mission (USA).gba
5C61A844,BMXE,08,3
```

I inserted the USA retail cart of course. The dumper takes the first "BMXE" entry in txt and set size to 16MB. All logical... but giving in an overdump for the real retail cart. Dunno if a couple of serial + header checksum would help here ?
